### ISSUE TITLE
Enable using llama.cpp's --draft <model> feature for speculative decoding

### ISSUE THREAD

**Issue by sammcj** [ollama/ollama#5800]
Howdy fine Ollama folks 👋 ,

Back this time last year llama.cpp added support for speculative decoding using a draft model parameter. https://github.com/ggerganov/llama.cpp/issues/2030

This can _massively_ speed up inference.

I was wondering if there's any chance you could look at adding the option for llama.cpp's `--model-draft` parameter that enables this?

---

It works by loading a smaller model (with the same tokeniser/family) in front of a larger model.

I know with exllamav2 you can get 100%-200% speed increases (seriously!) and the best part is - with no loss of quality.

For example, you might have:

- Main model: Qwen 72b Q4_K_M
- Draft model: Qwen 0.5b Q4_K_M

The result would be the memory usage of main Qwen 72b model + the tiny 0.5b draft model but around 4x the tokens/s you'd see with just the main model loaded.

I've been using exllamav2 instead of Ollama for this feature (and the 4bit K/V cache) recently and the performance really is astounding.

---

Parameters that can be passed to llama.cpp's server:

1. `--model-draft` (required) - the usage is the same as the existing `--model` used for loading normal models.
2. `--draft` (optional, but recommended to make available)
3. `--p-split` (optional, but recommended to make available)

There is also the following, but I think the default is probably fine 99% of the time:

- `--threads-draft`
- `--threads-batch-draft`

https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md?plain=1#L37

```
  -md,   --model-draft FNAME      draft model for speculative decoding (default: unused)
  -td,   --threads-draft N        number of threads to use during generation (default: same as --threads)
  -tbd,  --threads-batch-draft N  number of threads to use during batch and prompt processing (default: same as --threads-draft)
         --draft N                number of tokens to draft for speculative decoding (default: 5)
  -ps,   --p-split N              speculative decoding split probability (default: 0.1)
```

**Comment by sammcj** [comment]
I just found this old issue which while pretty out of date seems to be related - https://github.com/ollama/ollama/issues/1292

**Comment by jmorganca** [comment]
@sammcj this is really, really cool. Sorry for the late reply. By the way, check out https://infini-ai-lab.github.io/Sequoia-Page/ if you haven't yet.

**Comment by sammcj** [comment]
Oh that's a neat project too! thanks @jmorganca :)

**Comment by sammcj** [comment]
I could imagine using draft models with Ollama Modelfiles to be quite a nice combo, e.g:

```
FROM qwen2:72b-q4_k_m
DRAFT qwen2:1.5-q4_k_m
```

**Comment by enn-nafnlaus** [comment]
> I could imagine using draft models with Ollama Modelfiles to be quite a nice combo, e.g:
> 
> ```
> FROM qwen2:72b-q4_k_m
> DRAFT qwen2:1.5-q4_k_m
> ```

The best option is generally a wide-but-shallow sheared model rather than a generalist small model. For example, I made some GGUF conversions of the Yan et al models here:

https://huggingface.co/Nafnlaus/Wide-Sheared-LLaMA-290M-GGUF

Non-GGUFs here:

https://huggingface.co/minghaoyan

The TL/DR is that getting "an" answer out fast is more important than getting a good answer out.

**Comment by sammcj** [comment]
@enn-nafnlaus that's really interesting to see, I'll have to read the paper but I'm assuming this results in the draft model taking significantly less vRAM as it's essentially just the one layer being loaded (with the tokenizer)?

I suspect this might be what Exllamav2 does with it's draft model loading as they don't seem to use as much vRAM as loading it normally.

**Comment by sammcj** [comment]
I should add I find draft models/speculative decoding with Exllamav2 so useful that I often find myself choosing to use Ellamav2 (via [TabbyAPI](https://github.com/theroyallab/tabbyAPI)/[TabbyLoader](https://github.com/theroyallab/tabbyAPI-gradio-loader)) over Ollama when loading models larger than 30b~ - the performance improvements when running 70b models is nothing short of amazing.

**Comment by oxfighterjet** [comment]
Is this a feature the ollama maintainers would be interested in implementing? I'm asking because I'm considering giving it a shot.

Are there some pointers / suggestions for anyone looking into implementing this? Thank you.

**Comment by sammcj** [comment]
@oxfighterjet I think this would be amazing to add, combined with #6279 - this would bring Ollama up to speed with the likes of ExllamaV2 / TabbyAPI which have had these as core features for a long time.

I was actually planning on trying to get it merged after #6279 is merged. As such I'd be more than happy to work with you on this (just note I'm only a contributor - not a maintainer).

If you look at #6279 you'll see how I've added parameters that pass down to the underlying llama.cpp.

I would take the same approach but also make sure there is support for configuring the draft model in the Modelfile and API. This is something I did have in my PR (prior to the latest reactor for the new runners/server) but was asked to remove as Ollama didn't want to add new features to the API / CLI at the time, however for the draft model feature it will be required by design, I still have the code kicking round for this here: https://github.com/sammcj/ollama/pull/26/files.

Again - I was going to work on this after #6279 is merged, assuming it actually is merged in soon - I'd still be happy to do the work for this ticket, or work with you on it - be it doing a first pass for you to review/improve, or simply to help with peer review.

One thing I'd be aware of expectations wise, getting features merged into Ollama is _painfully_ slow, just to set your expectations up front 😅.

**Comment by oxfighterjet** [comment]
@sammcj Thank you for your helpful resources, they will most certainly come in handy! 
I see indeed that your #6279 PR has been a rollercoaster of a ride, I hope it can get merged soon.

I'll study the codebase and your PRs and see what I can contribute. I might get back to you with questions! :)

**Comment by bsu3338** [comment]
> @sammcj Thank you for your helpful resources, they will most certainly come in handy! I see indeed that your #6279 PR has been a rollercoaster of a ride, I hope it can get merged soon.
> 
> I'll study the codebase and your PRs and see what I can contribute. I might get back to you with questions! :)

Have you guys also considered the below approach? It looks like you could mix and match and 2 models. However, it might not be as performant.

https://huggingface.co/blog/universal_assisted_generation

**Comment by TheTerrasque** [comment]
I tried running llama server with speculative decoding to see if I could speed up some model, but I found out it's not supported by the server:

https://github.com/ggerganov/llama.cpp/issues/5877

**Comment by TheTerrasque** [comment]
https://github.com/ggerganov/llama.cpp/pull/10455 - this is now in llama.cpp server!

**Comment by oxfighterjet** [comment]
Great, I'm still on the ollama implementation and I'll be able to test it now. Will report back when I have a working prototype.

**Comment by chris-calo** [comment]
@oxfighterjet looks like #7875 was favoured over #7670, and is moving faster, if it helps any

**Comment by sammcj** [comment]
Looks like performance just got another big bump thanks to https://github.com/ggerganov/llama.cpp/pull/10586 (source: https://www.reddit.com/r/LocalLLaMA/comments/1h5uq43/llamacpp_bug_fixed_speculative_decoding_is_30/)

**Comment by cduk** [comment]
These options were tantalisingly mentioned in the opening post, but these don't seem to be valid options in llama-server. Have these been implemented in any branch or are these just proposals?

```
  -md,   --model-draft FNAME      draft model for speculative decoding (default: unused)
  -td,   --threads-draft N        number of threads to use during generation (default: same as --threads)
  -tbd,  --threads-batch-draft N  number of threads to use during batch and prompt processing (default: same as --threads-draft)
         --draft N                number of tokens to draft for speculative decoding (default: 5)
  -ps,   --p-split N              speculative decoding split probability (default: 0.1)
```

**Comment by cduk** [comment]
I will check again but I was referring specifically to flags -tbd, -td and -ps.

**Comment by bfroemel** [comment]
> I could imagine using draft models with Ollama Modelfiles to be quite a nice combo, e.g:
> 
> ```
> FROM qwen2:72b-q4_k_m
> DRAFT qwen2:1.5-q4_k_m
> ```
agreeing, the draft model (model-draft) and most related parameters (draft-max, draft-min, draft-p-min, ctx-size-draft) should be specified in the Modelfile. some parameters could be defined in the environment (device-draft, gpu-layers-draft) unless there is a good way to derive them automatically.

- model-draft: [Modelfile, DRAFT] ollama uses layers (media type: `application/vnd.ollama.image.model`) to reference model blobs. it would be nice to reuse the same mechanism and store the layer reference in the model manifest. Maybe just assume that the first `application/vnd.ollama.image.model` layer is the main model, and an optional additional `application/vnd.ollama.image.model` layer is the draft model?
- draft-max, draft-min, draft-p-min, ctx-size-draft [Modelfile, PARAMETER] Those parameters appear to be just Runner options, and can probably be very easily added.

**Comment by Steel-skull** [comment]
looks like https://github.com/ollama/ollama/pull/7875 was merged

**Comment by chris-calo** [comment]
@oxfighterjet are you still working on this?

**Comment by oxfighterjet** [comment]
@chris-calo yes.

**Comment by bfroemel** [comment]
just because it wasn't obvious to me: getting this into ollama is going to be more work than just handling down the mentioned parameters.

it appears that we basically have to replicate this as well:
https://github.com/ggerganov/llama.cpp/commit/9ca2e677626fce759d5d95c407c03677b9c87a26
and keeping track of fixes (for example, there are more):
https://github.com/ggerganov/llama.cpp/commit/84e1c33cde9e0a7aafcda2d4f21ba51c300482d7
https://github.com/ggerganov/llama.cpp/commit/1da7b765692764a8b33b08da61cbee63812a7bd9

**Comment by bfroemel** [comment]
Before moving forward with a prototype implementation it may be helpful to discuss the necessary changes?

Imo we roughly have the following tasks:
1. Parameter handling, how and where to define them, how to reference draft models such that the existing model repository can be used for loading a draft model, how to pass all required parameters down to a runner.
2. Reimplementation of the actual draft model feature in the runner, i.e., draft model loading (https://github.com/ollama/ollama/blob/cf4d7c52c47d753bd04a8791b9c6042271c40c1e/llama/runner/runner.go#L845) and using it during inference (https://github.com/ollama/ollama/blob/cf4d7c52c47d753bd04a8791b9c6042271c40c1e/llama/runner/runner.go#L360)
3. Reimplement or preferably somehow use existing utility code (cgo) located in llama.cpp (https://github.com/ggerganov/llama.cpp/blob/master/common/speculative.cpp, https://github.com/ggerganov/llama.cpp/blob/master/common/common.cpp) by appropriately extending https://github.com/ollama/ollama/blob/main/llama/llama.go

**Comment by mspinelli** [comment]
Maybe this is not helpful, but perhaps there are some additional ideas on how to easily add this functionality by looking at how the [llama-swap](https://github.com/mostlygeek/llama-swap/blob/main/examples/speculative-decoding/README.md) project accomplishes this?

**Comment by bfroemel** [comment]
Before having looked at the source I also assumed that ollama just starts llama.cpp server instances similar to llama-swap. I guess there are or have been good reasons why ollama reimplemented that part of llama.cpp; probably because of the added flexibility and maybe being able to implement some features quicker than having to wait for upstream. At least for this feature, upstream was faster; also it is my impression that llama.cpp server nowadays appears more sophisticated than what we have in ollama, so for the long run it might really be a good idea to look into adopting llama.cpp server directly as a runner (and add potentially missing instrumentation/control API to llama.cpp server).

Anyway, as I wanted to understand speculative decoding and getting into Go I tried to move forward with the previously outlined tasks and made progress with 1., and 3. (turned out ollama already made use of interfacing c++ code via C wrappers, so this was easy to extend). The second task is a bit of a struggle to debug. As soon as I have something of an initial proof of concept-quality solution to show in a couple of days and @oxfighterjet hasn't already done so, I'll open a PR...

**Comment by oxfighterjet** [comment]
@bfroemel Thanks for sharing your thoughts and your intentions. I am mostly interested in this feature being implemented at all, but my personal availability has decreased lately, with my work requiring more of my attention before the end of the year. It seems you are interested in taking over this issue and I'm glad to hand it over, I do not want to claim any exclusivity over it. If you have some ideas of how to implement it, please go ahead. I will anyway follow the progress of this issue closely, and am hoping for this feature to be propagated all the way to the top with open-webui :)

**Comment by bfroemel** [comment]
@oxfighterjet @sammcj Could you take a look at https://github.com/ollama/ollama/pull/8134 ? Testing/reviews/comments very welcome ;)

**Comment by sammcj** [comment]
Seeing more very positive things about the performance and surprisingly - TDP/power usage required with speculative decoding in llama.cpp: https://www.reddit.com/r/LocalLLaMA/comments/1hqlug2/revisting_llamacpp_speculative_decoding_w/

**Comment by zjh-nuc-AIOT** [comment]
现在这个功能实现了吗

**Comment by sammcj** [comment]
FYI LM Studio just added speculative decoding / draft model support in 0.3.10.

> We're excited to share LM Studio 0.3.10 (b1) in Beta with... 🥁Speculative Decoding!
>
> Speculative Decoding is a technique to gain inference speed ups, sometimes up to 1.5x-3x, using a combination of a "main model" and "draft model". This works best for a large main model and a very small draft model.                                                                                                                                                                                                                            
>
> How to turn on Speculative Decoding
>
> In the Chat or Server UI, you will see a new section titled "Speculative Decoding".     
>                                                                                                                               > Once you load a model, you will be able to choose a compatible draft model to be used        
>                                                                                                                          
> For API usage, pass the draft_model field in addition to the model field to pick which draft 
model to use

**Comment by dalisoft** [comment]
@sammcj Temporarily we see less performance improvements and later improvements could be more

**Comment by StevePierce** [comment]
Posting here since this seems to be the most active thread, just wanted to ask if speculative decoding is on the roadmap?

**Comment by hennas-waifson** [comment]
It would be so nice to have that feature. It would make a huge difference for many Ollama users.

**Comment by yurii-sio2** [comment]
My vote for this feature.

**Comment by iSevenDays** [comment]
I vote for this feature too!

**Comment by sammcj** [comment]
_Really_ big improvements from recent llama.cpp versions:

- **Llama.cpp** with draft model: **39.69Tk/s**
- **Ollama** without draft model: **29.75Tk/s**

FYI @jmorganca ^^

---

```
llama-server --port 12394 -ngl 99 --ctx-size 32768 -fa --cache-type-k q8_0 --cache-type-v q8_0 --host 0.0.0.0 --model-draft ./Qwen2.5-Coder-0.5B-Instruct-Q5_K_M.gguf --draft-max 24 --draft-min 1 --draft-p-min 0.6 --n-gpu-layers-draft 99 --parallel 4 --model ./Qwen2.5-Coder-32B-Instruct-Q5_K_M.gguf

prompt eval time =      90.59 ms /    73 tokens (    1.24 ms per token,   805.79 tokens per second)
       eval time =   20029.22 ms /   795 tokens (   25.19 ms per token,    39.69 tokens per second)
```

```
ollama run qwen2.5-coder-32b-instruct-128k:q5_k_m --verbose
>>> /set parameter num_ctx 32768

total duration:       27.74635847s
load duration:        8.672663ms
prompt eval count:    127 token(s)
prompt eval duration: 135.316513ms
prompt eval rate:     938.54 tokens/s
eval count:           821 token(s)
eval duration:        27.600339374s
eval rate:            29.75 tokens/s
```

**Comment by bfroemel** [comment]
Is this only https://github.com/ggml-org/llama.cpp/commit/abd4d0bc4f1a9a0e429bc8ee0d5ece2a394a0a39 or did you notice any other changes related to speculative decoding?

**Comment by sammcj** [comment]
@bfroemel unsure of the exact commit that made the difference but - damn, it's impressive - that's up there with ExllamaV2 speculative decoding performance now.

I'm seeing a solid **33.5%~ performance increase** by loading a tiny 0.5b draft model with hardly any additional vRAM usage and of course no quality difference.

**Comment by Abdulrahman392011** [comment]
you can always add the feature and make it dormant. apple use this strategy all the time. have it under development features and ask people to report anything related to this feature and explain that it's still in beta. 

this will give a bit of content on how it will perform as to not embarrass oneself if there is a system that is incompatible or crash or lower the performance. in cases like these it's better to have as much beta testers as possible and thankfully most of the ollama users are somewhat experienced with computers and can be proactive in testing such feature.

**Comment by pdevine** [comment]
I really want to do speculative decoding in Ollama, but my concern is always us trying to take too much on too quickly. Especially now that we have the new engine are we're slowly deprecating llama.cpp engine; if we add it in the new engine it would only work on a handful of models at first that work on the new engine (although maybe that's fine). I also want to make sure we figure out the local vs. hybrid story (i.e. offloading the big model to a different Ollama server).

**Comment by pdevine** [comment]
@sammcj hopefully you don't mind, but I changed the issue title since I think it's a broader topic. We wouldn't enable llama.cpp's draft mode since we're moving away from llama.cpp on the backend anyway.

**Comment by Abdulrahman392011** [comment]
well, no worries. whatever you guys are doing, keep doing it. the results are great. 

these things take patience and rushing it won't give us the results we need. so no pressure, after all we understand the ollama team is doing this cause they want to, not cause they need to or have to.

**Comment by sammcj** [comment]
@pdevine no worries at all! Not precious about the title by any means and am fully in support of any method of bringing speculative decoding to Ollama.

**Comment by Master-Pr0grammer** [comment]
@pdevine what is the reason behind wanting to move away from llama.cpp? would it not be more efficient to stick to llama.cpp, and instead of making your own engine to support features, just contribute to llama.cpp to implement your features? 

That way you get the added benefit of more community support.

Or has that been proven to difficult/inefficient?

**Comment by pdevine** [comment]
@Master-Pr0grammer I have utmost respect for ggml and the llama.cpp project and what Georgi has done, but we were finding that we were diverging too much from llama.cpp and our design philosophies are very different.

**Comment by Wladastic** [comment]
Instead of only adding this feature, why not allow users to split inferences in between layers?
Could even make a test script that goes through combinations of layers and stitch together a frankenmerge of the bigger and smaller llm.

**Comment by pdevine** [comment]
I have some ideas around how to get this going in the new engine. This hinges on getting the logprobs, but should be doable. Hopefully I'll have something more concrete details in a few weeks once I'm finished up with some other work.

**Comment by pdevine** [comment]
OK, I haven't forgotten about this, but we've been trying to get 0.10.0 out the door. We still need logprobs to be exposed properly to make it work.

**Comment by sammcj** [comment]
Thanks @pdevine , love your work!

**Comment by rpeinl** [comment]
There is a new GLM model version 4.5 out there in a bigger and smaller version similar to Llama4
https://huggingface.co/zai-org/GLM-4.5-Air
This looks very promising regarding model accuracy and it can do multi-token prediction (MTP).
Unfortunately, there is not much information available about how this works in the inference engine. However, there is a recent paper from Apple that links MTP to speculative decoding.
https://arxiv.org/html/2507.11851v1
Since tools like LMStudio already support GLM 4.5 and also supports speculative decoding, maybe it only works together.
Anyway, I would be extremely interested in getting this model to work in ollama, including MTP.

**Comment by BigArty** [comment]
Is it possible that there will be a way to make speculative decoding based on n-grams of some given text (or prompt and dialogue history)? It is by far the best way for weaker GPUs and similar or faster then 0.5B assistant models for ~8B - 14B generator models.

**Comment by BigArty** [comment]
@pdevine Are there any chance that this feature is still in development?

**Comment by dhirajlochib** [comment]
hi, ahm i've been working on implementing speculative decoding support and have completed the foundational infrastructure... here's the current status:

Implemented:
1. **Modelfile DRAFT Command** - Parse and store draft model references

FROM qwen2.5:3b
DRAFT qwen2.5:0.5b

2. **API & Config Support** - Added `Draft` field throughout the stack:
- `api.CreateRequest` and `api.ShowResponse`
- `types.ConfigV2` for persistence
- Model storage/retrieval in `server/create.go` and `server/images.go`

3. **Scheduler Integration** - Co-loading of draft and target models:
- `loadDraftModel()` for async background loading
- `GetLoadedRunner()` to retrieve loaded draft model

4. **Speculative Engine** (`speculative/speculative.go`):
- Draft token generation
- Batch verification with target model
- Acceptance criterion using rejection sampling (per Leviathan et al., 2022)
- Metrics tracking (acceptance rate, speedup estimation)

5. **Tests & Documentation** - Parser tests, unit tests, and Modelfile docs

what's not working yet

The actual **2-4x speedup doesn't activate** because the integration needs to go deeper into the runner's token generation loop. Currently:

- Draft model loads successfully 
- But `GenerateHandler` still uses standard single-model completion 
- The `SpeculativeCompletion` method exists but needs integration into the runner's core inference loop

Testing shows identical token generation rates with/without draft model because speculative decoding isn't engaging.

need guidance:
The final step requires changes to `runner/ollamarunner/runner.go` - specifically the token-by-token generation logic. This touches critical inference code that I'm less familiar with.

**Questions:**
1. Should the speculative logic live in the runner, or can it wrap the completion flow at a higher level?
2. Are there specific patterns in the runner for batched token verification I should follow?
3. Would the maintainers prefer to handle the runner integration, or should I continue working on it?

**Branch:** [`feature/speculative-decoding`](https://github.com/dhirajlochib/ollama/tree/feature/speculative-decoding)

Happy to continue working on this with guidance, or hand off the runner integration to someone more familiar with that codebase. The foundation is solid and ready for the final piece!!!

**Comment by Filipp-Druan** [comment]
Hello!
Please tell me what's going on with speculative decoding?
It's really important to me that this feature works. It's really hard without it! The models are incredibly slow!

Perhaps you could add Prompt Lookup Decoding? I really, really need fast program execution!

**Comment by pdevine** [comment]
OK, an update on this. Yes, I'm still looking at it, but I've been focusing on the MLX runner. I have a prototype of MTP working w/ MLX and a new multi-token sampler, but we need to get the new batching changes for the MLX runner in first which are also going to change the sampler before we can get this in.

Also, for non-Metal users, we're also working on getting the MLX runner to work on other platforms (i.e. CUDA), so there's a bit of juggling that needs to happen.

**Comment by alexander-potemkin** [comment]
> for non-Metal users, we're also working on getting the MLX runner to work on other platforms (i.e. CUDA), so there's a bit of juggling that needs to happen.

@pdevine , thanks for sharing! Does MLX runner has any benefits for non Mac?
I haven't heard anything on that, but it seems like there must be some, since you consider porting the code?

**Comment by Filipp-Druan** [comment]
@pdevine 
Excuse me, but what about speculative decoding based on n-grams? This is really, really, really important to me!
Llama.cpp already has this feature! You just need to add a command line option to Ollama! This can speed up inference significantly!

**Comment by ucffool** [comment]
meh. It can, but with MoE models doing some of the same heavy lifting built-in, that seems to be the current direction for speeding up inference.

**Comment by Filipp-Druan** [comment]
> meh. It can, but with MoE models doing some of the same heavy lifting built-in, that seems to be the current direction for speeding up inference.

That's not it. I need other models, not MoE.
I need to speed up the inference of regular models! You see, simple n-gram-based acceleration can improve speed at a very small cost!

**Comment by Filipp-Druan** [comment]
Using MoE reduces the model's capabilities compared to dense versions of the same size.
But n-grams don't!

**Comment by black-snow** [comment]
Coming back here after gemma4 mtp drafter announcement/release.
